libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/internal/navigation/NavigationHelper.kt

```diff
@@ -0,0 +1,184 @@
+ */
+object NavigationHelper {
+
+    private const val INDEX_ZERO = 0
+    private const val EMPTY_STRING = ""
+
+    /**
+     * When a milestones triggered, it's instruction needs to be built either using the provided
+     * string or an empty string.
+     */
+    fun buildInstructionString(
+        routeProgress: RouteProgress,
+        milestone: Milestone
+    ): String {
+        return milestone.instruction?.let {
+            milestone.instruction.buildInstruction(routeProgress)
+        } ?: EMPTY_STRING
+    }
+
+    /**
+     * Takes in the leg distance remaining value already calculated and if additional legs need to be
+     * traversed along after the current one, adds those distances and returns the new distance.
+     * Otherwise, if the route only contains one leg or the users on the last leg, this value will
+     * equal the leg distance remaining.
+     */
+    @JvmStatic
+    fun routeDistanceRemaining(
+        legDistanceRemaining: Double,
+        legIndex: Int,
+        directionsRoute: DirectionsRoute
+    ): Double {
+        var legDistanceRemaining = legDistanceRemaining
+        directionsRoute.legs()?.let { legs ->
+            val legsSize = legs.size
+            if (legsSize < 2) {
+                return legDistanceRemaining
+            }
+
+            for (i in legIndex + 1 until legsSize) {
+                legDistanceRemaining += legs[i].distance() ?: 0.0
+            }
+        }
+        return legDistanceRemaining
+    }
+
+    /**
+     * Given the current [DirectionsRoute] and leg / step index,
+     * return a list of [Point] representing the current step.
+     *
+     *
+     * This method is only used on a per-step basis as [PolylineUtils.decode]
+     * can be a heavy operation based on the length of the step.
+     *
+     *
+     * Returns null if index is invalid.
+     *
+     * @param directionsRoute for list of steps
+     * @param legIndex to get current step list
+     * @param stepIndex to get current step
+     * @return list of [Point] representing the current step
+     */
+
+    @JvmStatic
+    fun decodeStepPoints(
+        directionsRoute: DirectionsRoute,
+        currentPoints: List<Point>?,
+        legIndex: Int,
+        stepIndex: Int
+    ): List<Point>? {
+        val legs = directionsRoute.legs()
+        legs?.let { legs ->
+            if (legs.isEmpty()) return currentPoints
+            val steps = legs[legIndex].steps()
+            steps?.let { steps ->
+                if (steps.isEmpty()) return currentPoints
+                val invalidStepIndex = stepIndex < 0 || stepIndex > steps.size - 1
+                if (invalidStepIndex) return currentPoints
+                val step = steps[stepIndex]
+                step?.let { step ->
+                    val stepGeometry = step.geometry()
+                    stepGeometry?.let {
+                        return PolylineUtils.decode(stepGeometry, PRECISION_6)
+                    }
```

Comment:
Should we add return `currentPoints` in case if `stepGeometry` is `null` ??
```suggestion
                    } ?: return currentPoints
```

Reply:
Great catch! We can remove it from here and https://github.com/mapbox/mapbox-navigation-android/blob/c4e15b7220e322fa3cd83d5f63557d6c918c6a2b/libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/internal/navigation/NavigationHelper.java#L100 will be executed 👍